### DISCUSSION TITLE
Finit hangs after mounting filesystems from /etc/fstab

### DISCUSSION THREAD

**Discussion by ApartGorgon9849** [finit-project/finit#348] [Q&A] [answered]
I'm working on a custom distro, and I decided to use finit as the init system. I feel like I'm just missing something, but I'm not sure what. Shown below is an image of finit hanging. I have also attached my /etc/fstab, /etc/finit.conf, a log of the vm's entire lifecycle, and the config.h that was used to build finit. Also, if there are any additional files you'd like to see, I'd be happy to upload them. If you know what I'm missing, please let me know.
Thanks
![image](https://user-images.githubusercontent.com/60110285/214963021-675465e0-2bb8-4662-83df-9faad1596378.png)
[fstab.txt](https://github.com/troglobit/finit/files/10513882/fstab.txt)
[finit.conf.txt](https://github.com/troglobit/finit/files/10513885/finit.conf.txt)
[log.txt](https://github.com/troglobit/finit/files/10513892/log.txt)
[config.h.txt](https://github.com/troglobit/finit/files/10513895/config.h.txt)

**Comment by troglobit** [comment]
Since you are using yet unreleased software from the master branch, could you tell me which hash you are using?  (We had a regression recently that upended things for many.)

Enable finit.debug to get some more output, see <https://github.com/troglobit/finit/blob/master/doc/cmdline.md>.

**Reply by ApartGorgon9849** [reply]
Good Morning and thanks for your reply. git log tells me the hash of the commit I am using is 934a4db7d5f7a187f91ce5df084a2c19289628e6. Hope this helps. Thank you for your time.

EDIT: Forgot to upload log with finit.debug passed as a kernel parameter
[log.txt](https://github.com/troglobit/finit/files/10518309/log.txt)

**Reply by troglobit** [reply]
OK, so latest commit on master, good! Thanks for the log, I'll have a look at the log in detail over the weekend.

**Reply by ApartGorgon9849** [reply]
Thanks for the update. Would you want me to create an issue for this discussion?

**Reply by ApartGorgon9849** [reply]
I can also upload my rootfs in a couple of hours if it turns out I'm missing something in case that might help

**Reply by troglobit** [reply]
Let's make an issue when we isolate the problem. 

It would be great to have access to the rootfs and kernel if possible. That speeds up sending greatly. Thanks!

**Reply by ApartGorgon9849** [reply]
Okay. I will send the rootfs and the bzImage in about half an hour to an hour. Is there a different place you'd want me to upload the archive, or is here in this thread fine?

**Reply by troglobit** [reply]
If you can dump it all in this thread, that would be great. I don't have any I other storage to offer, unfortunately. 

I won't be able to look at it until tomorrow though, so no panic.

**Reply by troglobit** [reply]
Oh, would be great if you could detail your qemu line, there are quite a few possibilities, which have tripped up many before.

**Reply by ApartGorgon9849** [reply]
Apologies it took so long. I had gotten caught up in a few things, and github wouldn't let me upload the rootfs, since it was over 25mb (in actuality, it is 5gb compressed, and 16gb uncompressed). My qemu launch command is as follows: ``
qemu-system-x86_64 -nographic -kernel bzImage -drive file=rootfs.img -m 1024 -append "root=/dev/sda console=ttyS0 init=/sbin/finit" |& tee log.txt
``. Hope this helps. Thank you again
[Google Drive - rootfs](https://drive.google.com/file/d/1rYMkSJku_oE1L1fjgiL_rNWy-84Xg3wb/view?usp=sharing)
[bzImage.zip](https://github.com/troglobit/finit/files/10525279/bzImage.zip)

**Reply by troglobit** [reply]
Thanks, this made it really easy to reproduce!

I mounted the rootfs and added this to your `/etc/finit.conf`:

    tty [12345789] console noclear nologin

Starting up again I now had a root shell and could probe it further.  Calling `initctl` showed acpid as waiting on `syslogd`, so I modified `finit.conf` as follows and called `initctl reload`:

    runlevel 2
    log size=100k count=4
    
    service [S12345] env:-/etc/conf.d/syslog syslogd -n            -- Syslog daemon
    service [S12345] <pid/syslogd> env:-/etc/conf.d/klogd klogd -n -- Kernel log daemon
    service [2345] <pid/syslogd> log acpid -f                      -- ACPI daemon        
    tty [12345789] console noclear nologin

Now both `acpid` and `klogd` failed to start.  So I started poking around a bit and realized that a lot of what I usually expect to be available on a Finit system was missing:

  - `/lib/plugins/*`
  - `/etc/finit.d/*`

The `finit.d/` directory is not mandatory, Finit works fine with just `finit.conf`.  However, to get the dependency handling to work, the `pidfile.so` plugin must be loaded, otherwise no pid conditions can be asserted.  You can of course use the bare minimum of Finit, without any plugins, but if you try to start a system without synchronization then you're a bit on your own.

I must say I'm not really sure what you expected would happen?  I'm sorry to say this, but have you looked at the `contrib/` section in the repository or read any of the documentation?  For reference you can have a look at some of the examples build around [Buildroot](https://buildroot.org) that I've published:

 - https://github.com/troglobit/myLinux
 - https://github.com/troglobit/br2-finit-demo, which uses:
   - https://github.com/troglobit/br2-ext-finit
   - https://github.com/troglobit/finit-skel

I really recommend the [br2-finit-demo](https://github.com/troglobit/br2-finit-demo) it's fairly straight forward to build and get with running.  Having a working reference when setting up your own system is always handy.  A walk-through is available here: https://troglobit.com/post/2022-12-26-buildroot-demo-of-finit/

Good Luck!

> Tip: add `-M q35,accel=kvm -cpu host -enable-kvm` to your qemu-system line to speed it up a bit.

**Reply by ApartGorgon9849** [reply]
Thank you for your time and assistance. I had looked through the documentation and contrib folder, but there was no mention of exactly what was required. I'll poke through the resources you posted. Thank you again

EDIT: Glad it was user error and not a bug

EDIT 2: I did enable build-static with configure, and I'm curious if that caused pidfile.so to not be built?

**Comment by troglobit** [comment, marked as answer]
I've tried to make the build instructions as clear as possible, but I guess there's always room for improvement.  

Hmm, yeah you're right. When building a static Finit the pidfile plugin should be built-in.  Sorry, but the static build should probably be marked as deprecated, and then removed in v5.  I've never used it myself, and was pushed by a very vocal user 5-10 years back.  I should know better than try to support use-cases I have no way of testing myself 😞 

I recommend going with a dynamic build, doing a proper "make install" and setting up your system .conf file(s) with help from the contrib section.
